### PR TITLE
EVA-2485 : Service for reading and storing counts from ingestion pipelines

### PR DESCRIPTION
**Pull request by nitin-ebi**

### CODE REVIEW

**Review by tcezard — COMMENTED** [review]
More competent people can comment on the code but I was wondering if this code should be sitting in [eva-ws](https://github.com/EBIvariation/eva-ws)

**Review by sundarvenkata-EBI — COMMENTED** [review]
Looks good! Will approve after DEV meeting discussion on additional libraries being used (Lombok, Jupiter, Jersey etc.,)

File: count-stats/pom.xml [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+            <scope>runtime</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-test</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter</artifactId>
```
Comment (andresfsilva) [inline count-stats/pom.xml:40]: `spring-boot-starter-test` is already a dependency and it includes `junit` which is what we are using across other repositories. I don't think this is needed unless there is a reason for switching to `junit jupiter`

File: count-stats/pom.xml [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    <packaging>war</packaging>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-data-jdbc</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-jersey</artifactId>
```
Comment (andresfsilva) [inline count-stats/pom.xml:22]: I don't think this is being used

File: count-stats/src/main/java/uk/ac/ebi/eva/countstats/model/CountDto.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package uk.ac.ebi.eva.countstats.model;
+
+import lombok.Value;
+
+import java.math.BigInteger;
+
+@Value
+public class CountDto {
```
Comment (andresfsilva) [inline count-stats/src/main/java/uk/ac/ebi/eva/countstats/model/CountDto.java:8]: I think this is part of the model and should not be named `Dto`

File: count-stats/pom.xml [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter</artifactId>
+            <version>RELEASE</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
```
Comment (andresfsilva) [inline count-stats/pom.xml:46]: I actually like `lombok`. We have not used it but we can discuss it in the dev meeting

File: count-stats/src/main/java/uk/ac/ebi/eva/countstats/service/CountService.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import uk.ac.ebi.eva.countstats.model.CountDto;
+import uk.ac.ebi.eva.countstats.repository.CountRepository;
+
+import java.util.List;
+
+@Service
+public class CountService {
+    @Autowired
```
Comment (andresfsilva) [inline count-stats/src/main/java/uk/ac/ebi/eva/countstats/service/CountService.java:12]: My personal preference is to use [constructor injection](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#using.spring-beans-and-dependency-injection), although I know we use field injection in many repositories.

File: count-stats/src/test/java/uk/ac/ebi/eva/countstats/CountStatsApplicationTests.java [hunk 6]
```diff
@@ -0,0 +1,13 @@
+package uk.ac.ebi.eva.countstats;
+
+import org.junit.jupiter.api.Test;
+import org.springframework.boot.test.context.SpringBootTest;
+
+@SpringBootTest
+class CountStatsApplicationTests {
```
Comment (andresfsilva) [inline count-stats/src/test/java/uk/ac/ebi/eva/countstats/CountStatsApplicationTests.java:7]: Are tests part of another ticket?

**Review by andresfsilva — CHANGES_REQUESTED** [review]
I agree with @tcezard that it makes sense to include this API as part of [eva-ws](https://github.com/EBIvariation/eva-ws)

File: count-stats/src/main/java/uk/ac/ebi/eva/countstats/service/CountService.java (hunk 5 above)
Comment (nitin-ebi) [reply count-stats/src/main/java/uk/ac/ebi/eva/countstats/service/CountService.java:12]: Removed field injection and added constructor injection in both Controller and Service

File: count-stats/pom.xml (hunk 4 above)
Comment (nitin-ebi) [reply count-stats/pom.xml:46]: Yes, removes a lot of boiler pot code. Let's discuss in dev meeting

File: count-stats/src/main/java/uk/ac/ebi/eva/countstats/model/CountDto.java (hunk 3 above)
Comment (nitin-ebi) [reply count-stats/src/main/java/uk/ac/ebi/eva/countstats/model/CountDto.java:8]: renamed to count

File: count-stats/pom.xml (hunk 2 above)
Comment (nitin-ebi) [reply count-stats/pom.xml:22]: Removed. 
Was using it to return response from controller, but yes not required unless we have something that spring cannot provided by default

File: count-stats/pom.xml (hunk 1 above)
Comment (nitin-ebi) [reply count-stats/pom.xml:40]: Removed.
This is for Junit 5. Spring boot by default uses Junit 4, but I guess that is fine for our use case.

File: count-stats/src/main/java/uk/ac/ebi/eva/countstats/controller/CountController.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+import javax.ws.rs.core.Response;
+import java.util.List;
+
+@RestController
+@RequestMapping("/v1/countstats")
+public class CountController {
+    @Autowired
+    private CountService countService;
+
+    @GetMapping("status")
```
[the remark this answers is not in the document]
Comment (nitin-ebi) [reply count-stats/src/main/java/uk/ac/ebi/eva/countstats/controller/CountController.java:17]: Removed and added actuator
was just testing it with this. forget to remove.

File: count-stats/src/main/java/uk/ac/ebi/eva/countstats/repository/CountRepository.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+import org.springframework.jdbc.core.namedparam.SqlParameterSource;
+import org.springframework.stereotype.Repository;
+import uk.ac.ebi.eva.countstats.model.CountDto;
+
+import java.math.BigInteger;
+import java.sql.Types;
+import java.util.List;
+
+@Repository
+public class CountRepository {
```
[the remark this answers is not in the document]
Comment (nitin-ebi) [reply count-stats/src/main/java/uk/ac/ebi/eva/countstats/repository/CountRepository.java]: JPA does not have out of the box support for PostgresSql (jsonb).  
I have changed the code to use JPA and Entity, but for the queries we will mostly have to rely on writing native queries. Specifically where we are going to query jsonb column (which I guess is going to be the case in most of the queries)

File: count-stats/src/test/java/uk/ac/ebi/eva/countstats/CountStatsApplicationTests.java (hunk 6 above)
Comment (nitin-ebi) [reply count-stats/src/test/java/uk/ac/ebi/eva/countstats/CountStatsApplicationTests.java:7]: Added Test. 
Have added test containers for integration test. Again, this is something we can discuss in our dev meeting.

**Review by andresfsilva — APPROVED** [review]

File: count-stats/pom.xml [hunk 9]
```diff
@@ -36,9 +36,15 @@
...
         <dependency>
-            <groupId>org.junit.jupiter</groupId>
+            <groupId>org.testcontainers</groupId>
+            <artifactId>postgresql</artifactId>
+            <version>1.15.3</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.testcontainers</groupId>
             <artifactId>junit-jupiter</artifactId>
```
Comment (sundarvenkata-EBI) [inline count-stats/pom.xml:46]: I think we should explicitly exclude the JUnit4 package that ships with Spring if we want to include this.

**Review by sundarvenkata-EBI — APPROVED** [review]

File: count-stats/pom.xml (hunk 9 above)
Comment (nitin-ebi) [reply count-stats/pom.xml:46]: Current version of Testcontainers depends on Junit-4. [See this issue](https://github.com/testcontainers/testcontainers-java/issues/970). 
They have a plan to remove this dependency in Testcontainers 2.0.
Till that time, we can explore other options like the one given [here](https://github.com/testcontainers/testcontainers-java/issues/970#issuecomment-625044008)

**Comment by nitin-ebi** [comment]
Closing this PR as this code needs to be in EVA-WS and not in EVA-Tools. 
Will raise a new PR for EVA-WS

**Pull request closed without merging**